seed/views/cycles.py

```diff
@@ -4,60 +4,187 @@
 :copyright (c) 2014 - 2016, The Regents of the University of California, through Lawrence Berkeley National Laboratory (subject to receipt of any required approvals from the U.S. Department of Energy) and contributors. All rights reserved.  # NOQA
 :author Paul Munday<paul@paulmunday.net>
 """
+import datetime
+
+from django.forms.models import model_to_dict
+from django.http import JsonResponse
 from rest_framework import status
 from rest_framework.renderers import JSONRenderer
-from rest_framework.response import Response
-from rest_framework.viewsets import ModelViewSet
+from rest_framework.viewsets import GenericViewSet
 
-from seed.decorators import DecoratorMixin
-from seed.models import Cycle
+from seed.decorators import ajax_request_class
+from seed.lib.superperms.orgs.decorators import has_perm_class
+from seed.lib.superperms.orgs.models import Organization
+from seed.models import Cycle, PropertyView, TaxLotView
 from seed.serializers.cycles import CycleSerializer
-from seed.utils.api import drf_api_endpoint
+from seed.utils.api import api_endpoint_class
 
 
-class CycleView(DecoratorMixin(drf_api_endpoint), ModelViewSet):
+class CycleView(GenericViewSet):
     renderer_classes = (JSONRenderer,)
     serializer_class = CycleSerializer
 
-    def get_queryset(self):
-        return Cycle.objects.filter(
-            organization_id=self.request.GET['organization_id']
-        ).order_by('name')
-
-    def get_object(self):
-        """Override this so we can easily use the UpdateModelMixin."""
-        # self.cycle_pk set in update_cycle()
-        pk = getattr(self, 'cycle_pk', None)
-        return Cycle.objects.get(pk=pk)
-
-    def list_cycle(self, request, *args, **kwargs):
-        qs = self.get_queryset()
-        if not qs:
-            result = {'status': 'error', 'message': 'No cycles found.'}
-            status_code = status.HTTP_404_NOT_FOUND
+    @api_endpoint_class
+    @ajax_request_class
+    def retrieve(self, request, pk=None):
+        """
+        Retrieves a cycle
+        ---
+        parameters:
+            - name: organization_id
+              description: The organization_id for this user's organization
+              required: true
+              paramType: query
+        """
+        # make sure query org id is in this user's orgs
+        org_id_in_query = request.query_params.get('organization_id', None)
+        cycles = Cycle.objects.filter(
+            organization_id=org_id_in_query, pk=pk
+        )
+        if cycles.exists():
+            cycle = cycles[0]
+        else:
+            return JsonResponse({
+                'status': 'error',
+                'message': 'Could not access cycle with id={}'.format(pk)
+            }, status=status.HTTP_403_FORBIDDEN)
```

Comment:
nvm: Receiving a 403 response is the server telling you, “I’m sorry. I know who you are–I believe who you say you are–but you just don’t have permission to access this resource. Maybe if you ask the system administrator nicely, you’ll get permission. But please don’t bother me again until your predicament changes.”